lib/rebaser-core/src/api_types.rs

```diff
@@ -0,0 +1,483 @@
+
+        assert!(!CoolRequest::is_message_version_supported(0));
+        assert!(!CoolRequest::is_message_version_supported(5));
+        assert!(!CoolRequest::is_message_version_supported(42));
+
+        let req_old = CoolRequestVersions::V2(CoolRequestV2 {
+            id: RequestId::new(),
+        });
+        dbg!(&req_old);
+
+        let mut req = req_old.into_current_version().expect("failed to upgrade");
+        dbg!(&req);
+
+        dbg!(&req.organization);
+        req.organization = Some("acme".to_string());
+        dbg!(&req.organization);
+
+        dbg!(&req);
+
+        publish(req).expect("failed to publish");
+    }
+
+    mod v1 {
+        use serde::{Deserialize, Serialize};
+
+        use super::RequestId;
+
+        #[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
+        #[serde(rename_all = "camelCase")]
+        pub struct CoolRequestV1 {
+            pub id: RequestId,
+        }
+    }
+
+    mod v2 {
+        use serde::{Deserialize, Serialize};
+
+        use super::{v1::CoolRequestV1, RequestId, UpgradeError};
+
+        #[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
+        #[serde(rename_all = "camelCase")]
+        pub struct CoolRequestV2 {
+            pub id: RequestId,
+        }
+
+        impl TryFrom<CoolRequestV1> for CoolRequestV2 {
+            type Error = UpgradeError;
+
+            fn try_from(value: CoolRequestV1) -> Result<Self, Self::Error> {
+                Ok(Self { id: value.id })
+            }
+        }
+    }
+
+    mod v3 {
+        use serde::{Deserialize, Serialize};
+
+        use super::{v2::CoolRequestV2, RequestId, UpgradeError};
+
+        #[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
+        #[serde(rename_all = "camelCase")]
+        pub struct CoolRequestV3 {
+            pub id: RequestId,
+            pub name: String,
+        }
+
+        impl TryFrom<CoolRequestV2> for CoolRequestV3 {
+            type Error = UpgradeError;
+
+            fn try_from(value: CoolRequestV2) -> Result<Self, Self::Error> {
+                Ok(Self {
+                    id: value.id,
+                    name: "<unknown>".to_string(),
+                })
+            }
+        }
+    }
+
+    mod v4 {
+        use serde::{Deserialize, Serialize};
+
+        use super::{v3::CoolRequestV3, RequestId, UpgradeError};
+
+        #[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
+        #[serde(rename_all = "camelCase")]
+        pub struct CoolRequestV4 {
+            pub id: RequestId,
+            pub name: String,
+            pub organization: Option<String>,
+        }
+
+        impl TryFrom<CoolRequestV3> for CoolRequestV4 {
+            type Error = UpgradeError;
+
+            fn try_from(value: CoolRequestV3) -> Result<Self, Self::Error> {
+                Ok(Self {
+                    id: value.id,
+                    name: value.name,
+                    organization: None,
+                })
+            }
```

Comment:
This is a really clever idea for upgrade testing before we even use or need it. Well done